### PR TITLE
Add display-name-property rule

### PR DESCRIPTION
**Pull request by andrewlor**
https://github.com/Expensify/App/issues/7822

`display-name-property` is a rule that will error when there is a react view file component that is either:
- A class component that sets the displayName property
- A functional component that does not set the displayName property 

Overview of how the code works: we only look at react view files using the helper `isReactViewFile`, which uses the path of the file. We then infer the component name from the file name, whether that be `Component.js` or `Component/index.js`. We then search the root level of the program tree in that file for a `VariableDeclaration` or `ClassDeclaration` that matches the component name to determine if it is a class or functional component. We can then finally search for the existence of a displayName property assignment. Based on that we have enough information to decide to report an error or not.

### Examples in the codebase

Screen Shot 2022-08-06 at 11 15 29 AM

Screen Shot 2022-08-06 at 11 15 48 AM

### CODE REVIEW

**Comment by andrewlor** [comment]
@MonilBhavsar @rushatgabhane Sorry I'm not sure if you guys missed this because I was not able to add you guys as reviewers to the PR. Have a look and let me know if there is any feedback. Thanks.

File: eslint-plugin-expensify/display-name-property.js [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    create: context => ({
+        Program(node) {
+            // Only looking at react files
+            if (!isReactViewFile(context.getFilename())) {
+                return;
+            }
+
+            // Infer the component name from the filename
+            const filenameParts = _.split(context.getFilename(), '/');
+            let componentName = _(filenameParts).last().split('.')[0];
```
[the remark this answers is not in the document]
Comment (andrewlor) [reply eslint-plugin-expensify/display-name-property.js:15]: Yes I believe so. Here is an example where I added `displayName` to a class component in an `index.android.js` file.

Screen Shot 2022-08-15 at 10 21 12 PM

**Comment by andrewlor** [comment]
For interest, here is a complete list of lint errors I see on the `main` branch right now when I add this rule.

```
/Users/andrewlor/dev/Expensify/src/components/AddPaymentMethodMenu.js
  1:1  error  Function components must set the displayName property  rulesdir/display-name-property

/Users/andrewlor/dev/Expensify/src/components/AddressSearch.js
  1:1  error  Function components must set the displayName property  rulesdir/display-name-property

/Users/andrewlor/dev/Expensify/src/components/CheckboxWithLabel.js
  1:1  error  Class components must not set the displayName property  rulesdir/display-name-property

/Users/andrewlor/dev/Expensify/src/components/FAB/fabPropTypes.js
  1:1  error  Function components must set the displayName property  rulesdir/display-name-property

/Users/andrewlor/dev/Expensify/src/components/FormScrollView.js
  1:1  error  Function components must set the displayName property  rulesdir/display-name-property

/Users/andrewlor/dev/Expensify/src/components/HTMLEngineProvider/HTMLRenderers/PreRenderer/index.js
  1:1  error  Class components must not set the displayName property  rulesdir/display-name-property

/Users/andrewlor/dev/Expensify/src/components/HeaderWithCloseButton.js
  1:1  error  Class components must not set the displayName property  rulesdir/display-name-property

/Users/andrewlor/dev/Expensify/src/components/InlineCodeBlock/inlineCodeBlockPropTypes.js
  1:1  error  Function components must set the displayName property  rulesdir/display-name-property

/Users/andrewlor/dev/Expensify/src/components/KeyboardSpacer/index.js
  7:1  error  Function components must set the displayName property  rulesdir/display-name-property

/Users/andrewlor/dev/Expensify/src/components/MultipleAvatars.js
  1:1  error  Function components must set the displayName property  rulesdir/display-name-property

/Users/andrewlor/dev/Expensify/src/components/PDFView/index.js
  1:1  error  Class components must not set the displayName property  rulesdir/display-name-property

/Users/andrewlor/dev/Expensify/src/components/PlaidLink/index.native.js
  1:1  error  Function components must set the displayName property  rulesdir/display-name-property

/Users/andrewlor/dev/Expensify/src/components/PlaidLink/plaidLinkPropTypes.js
  1:1  error  Function components must set the displayName property  rulesdir/display-name-property

/Users/andrewlor/dev/Expensify/src/components/SVGImage/propTypes.js
  1:1  error  Function components must set the displayName property  rulesdir/display-name-property

/Users/andrewlor/dev/Expensify/src/components/SafeArea/index.ios.js
  1:1  error  Function components must set the displayName property  rulesdir/display-name-property

/Users/andrewlor/dev/Expensify/src/components/ScreenWrapper/index.android.js
  1:1  error  Function components must set the displayName property  rulesdir/display-name-property

/Users/andrewlor/dev/Expensify/src/components/ScreenWrapper/index.js
  1:1  error  Function components must set the displayName property  rulesdir/display-name-property

/Users/andrewlor/dev/Expensify/src/components/ScreenWrapper/propTypes.js
  1:1  error  Function components must set the displayName property  rulesdir/display-name-property

/Users/andrewlor/dev/Expensify/src/components/TestToolMenu.js
  1:1  error  Function components must set the displayName property  rulesdir/display-name-property

/Users/andrewlor/dev/Expensify/src/components/TestToolRow.js
  1:1  error  Function components must set the displayName property  rulesdir/display-name-property

/Users/andrewlor/dev/Expensify/src/components/ThreeDotsMenu/index.js
  1:1  error  Class components must not set the displayName property  rulesdir/display-name-property

/Users/andrewlor/dev/Expensify/src/components/WalletStatementModal/index.js
  1:1  error  Class components must not set the displayName property  rulesdir/display-name-property

/Users/andrewlor/dev/Expensify/src/components/WalletStatementModal/index.native.js
  1:1  error  Class components must not set the displayName property  rulesdir/display-name-property

/Users/andrewlor/dev/Expensify/src/pages/AddPersonalBankAccountPage.js
  1:1  error  Class components must not set the displayName property  rulesdir/display-name-property

/Users/andrewlor/dev/Expensify/src/pages/ReportSettingsPage.js
  1:1  error  Class components must not set the displayName property  rulesdir/display-name-property

/Users/andrewlor/dev/Expensify/src/pages/home/sidebar/SidebarScreen/sidebarPropTypes.js
  1:1  error  Function components must set the displayName property  rulesdir/display-name-property

/Users/andrewlor/dev/Expensify/src/pages/iou/steps/IOUParticipantsPage/IOUParticipantsSplit.js
  1:1  error  Class components must not set the displayName property  rulesdir/display-name-property

/Users/andrewlor/dev/Expensify/src/pages/safeAreaInsetPropTypes.js
  1:1  error  Function components must set the displayName property  rulesdir/display-name-property

/Users/andrewlor/dev/Expensify/src/pages/settings/Payments/AddPayPalMePage.js
  1:1  error  Class components must not set the displayName property  rulesdir/display-name-property

/Users/andrewlor/dev/Expensify/src/pages/settings/Payments/walletTransferPropTypes.js
  1:1  error  Function components must set the displayName property  rulesdir/display-name-property

/Users/andrewlor/dev/Expensify/src/pages/settings/Profile/ProfilePage.js
  1:1  error  Class components must not set the displayName property  rulesdir/display-name-property

/Users/andrewlor/dev/Expensify/src/pages/settings/Security/CloseAccountPage.js
  1:1  error  Class components must not set the displayName property  rulesdir/display-name-property

/Users/andrewlor/dev/Expensify/src/pages/wallet/WalletStatementPage.js
  1:1  error  Class components must not set the displayName property  rulesdir/display-name-property

/Users/andrewlor/dev/Expensify/src/pages/workspace/WorkspaceInitialPage.js
  1:1  error  Class components must not set the displayName property  rulesdir/display-name-property

/Users/andrewlor/dev/Expensify/src/pages/workspace/reimburse/WorkspaceReimburseView.js
  1:1  error  Class components must not set the displayName property  rulesdir/display-name-property

✖ 35 problems (35 errors, 0 warnings)
```

**Comment by andrewlor** [comment]
Actually wait, I'll have to make one more change to the rule. I just noticed it is reporting errors for propType files. Will have to find a way to ignore that case.

File: eslint-plugin-expensify/display-name-property.js [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+            // Find a root level assignment expression that matches <componentName>.displayName = ...
+            const hasDisplayNameProperty = _.find(node.body, bodyNode => bodyNode.type === 'ExpressionStatement'
+                && _.get(bodyNode, 'expression.type', '') === 'AssignmentExpression'
+                && _.get(bodyNode, 'expression.left.property.name', '') === 'displayName'
+                && _.get(bodyNode, 'expression.left.object.name', '') === componentName);
+
+            // Allow function components with displayName set
+            if (isFunctionComponent && hasDisplayNameProperty) {
+                return;
```
[the remark this answers is not in the document]
Comment (andrewlor) [reply eslint-plugin-expensify/display-name-property.js:41]: Hmm I can look into it.

File: package.json [hunk 3]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "eslint-config-expensify",
-  "version": "2.0.29",
+  "version": "2.0.30",
```
[the remark this answers is not in the document]
Comment (MonilBhavsar) [reply package.json:3]: yes!

File: eslint-plugin-expensify/display-name-property.js [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+
+            if (isFunctionComponent) {
+                // Allow function components with displayName set
+                if (hasDisplayNameProperty) {
+                    return;
+                }
+
+                // Ignore noop components, ie. const Test = () => null;
+                const component = isFunctionComponent;
+                const isNoopComponent = _.get(component, 'declarations[0].init.type') === 'ArrowFunctionExpression'
```
[the remark this answers is not in the document]
Comment (andrewlor) [reply eslint-plugin-expensify/display-name-property.js:47]: isNoopComponent is actually a boolean, so I'll leave it.

**Review by MonilBhavsar — APPROVED** [review]
Looks good to me. Before I merge, a quick question - There will be another PR in App repo to fix the lint failures, right?

**Comment by MonilBhavsar** [comment]
@rushatgabhane package is updated to 2.0.30 https://www.npmjs.com/package/eslint-config-expensify
Feel free to open a PR

**Pull request merged**